src/optimagic/deprecations.py

```diff
@@ -421,3 +459,39 @@ def replace_and_warn_about_deprecated_derivatives(candidate, name):
             out.append(key_to_marker[key](func))
 
     return out
+
+
+def pre_process_constraints(
```

Comment:
This will need a docstring that explains why we convert everything to dict now and that we should go the opposite way once we have time to refactor the internals.